Bluetooth: Gatt: Subscription callback

asbjornsabo opened the pull request:
Create separate callback for CCC subscriptions, to be able to return the subscription parameters.

This is an API change, discussed in https://github.com/zephyrproject-rtos/zephyr/issues/45237

asbjornsabo committed:
Bluetooth: Gatt: Replace callback in gatt subscription callback.

Add a "subscribe" callback in the bt_gatt_subscribe params, and
deprecate the "write" callback.

The purpose of this is to be able to return the subscription
parameters in the callback.  The write callback
- (in principle) returns write parameters
- in fact returns nothing - the pointer is set to NULL

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Gatt: Update users of write callback in subscription params

Update the users of the "write" callback in the
bt_gatt_subscribe_params to use the new "subscribe" callback instead.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1618,6 +1618,18 @@ typedef uint8_t (*bt_gatt_notify_func_t)(struct bt_conn *conn,
+/** @typedef bt_gatt_subscribe_func_t
+ *  @brief Subscription callback function
+ *
+ * If given, called with the subscription parameters given when subscribing
```
```suggestion
 *  If given, called with the subscription parameters given when subscribing
```

asbjornsabo replied:
Fixed

Thalley reviewed (commented):
This was exactly what I had in mind too for the solution :)

asbjornsabo commented:
> This was exactly what I had in mind too for the solution :)

Well, it was more or less your idea, wan't it? :-)

asbjornsabo commented:
The `write`function has been deprecated (by this PR), but is still used (=kept) in the implementation to provide backwards compatibility.
But the use of the deprecated function causes the build to fail.
How do I work around this?

Thalley commented:
> The `write`function has been deprecated (by this PR), but is still used (=kept) in the implementation to provide backwards compatibility. But the use of the deprecated function causes the build to fail. How do I work around this?

We can't use `__deprecated` for callback unfortunately :( Went through the same issue with https://github.com/zephyrproject-rtos/zephyr/issues/43543 and simply had to remove them.

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1618,6 +1618,18 @@ typedef uint8_t (*bt_gatt_notify_func_t)(struct bt_conn *conn,
+/** @typedef bt_gatt_subscribe_func_t
+ *  @brief Subscription callback function
+ *
+ *  If given, called with the subscription parameters given when subscribing
```
It feels like this line should be for `bt_gatt_subscribe_params.subscribe`, and not the in the description of the typedef

asbjornsabo replied:
Agree

asbjornsabo replied:
Fixed

Thalley commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1660,6 +1672,9 @@ struct bt_gatt_subscribe_params {
 	/** Subscribe CCC write request response callback */
+	bt_gatt_subscribe_func_t subscribe;
+
+	/** DEPRECATED subscribe CCC write response callback */
```
Since we cannot properly deprecate it, is this the right way to do it @carlescufi ? Shouldn't we just remove the callback?

carlescufi replied:
Yes, a stable API change means that, changing a stable API. So if we are not changing the existing `write` callback then we should remove it from here.

carlescufi replied:
@asbjornsabo why are we not changing the signature of `bt_gatt_write_func_t` instead? is it because it's used in other places?

carlescufi replied:
Understood, thank you @asbjornsabo.
@Thalley, since you added that callback, would you mind ACKing this change here?

Thalley replied:
I am in favor of replacing the `write` callback with the new callback completely (i.e. remove `write` and add `subscribe`, without deprecation).

asbjornsabo replied:
Dsicussed in discord - keep deprecated callback, as that is cheap.

carlescufi commented on doc/releases/release-notes-3.1.rst:
```diff
@@ -78,6 +78,9 @@ Bluetooth
 
   * The :c:macro:bt_conn_index function now takes a `const struct bt_conn`.
 
+  * The `struct bt_gatt_subscribe_params` :c:func:`write` callback
```
```suggestion
  * In the `struct bt_gatt_subscribe_params`, the  :c:func:`write` callback
```

carlescufi commented:
API meeting:

- No objections

carlescufi commented:
@jhedberg could you please take a look and let us know if you'd prefer deprecating instead of using the Stable API Change procedure?

jhedberg commented:
> @jhedberg could you please take a look and let us know if you'd prefer deprecating instead of using the Stable API Change procedure?

The overhead from maintaining support for the deprecated callback seems to be only a few lines of code, so I'm fine with taking that approach.

asbjornsabo commented on include/zephyr/bluetooth/gatt.h:
```diff
@@ -1660,6 +1672,9 @@ struct bt_gatt_subscribe_params {
 	/** Subscribe CCC write request response callback */
+	bt_gatt_subscribe_func_t subscribe;
+
+	/** DEPRECATED subscribe CCC write response callback */
```
Fixed

asbjornsabo commented:
Rebased, and fixed the doxygen comment from @gmarull 

Unsure about what the conclusion was above on whether to remove or deprecate the previous callback.  I suspect no-one is using it, so should be fine to remove.  But have gone with @jhedberg now and kept it as deprecated.

Will take out of draft as soon as I have tried it as base for the change I need it for.

asbjornsabo commented:
Pushed again - moved documentation from function type to struct, as suggested by Emil.  (Was hidden within a comment that had been marked as resolved.)

Thalley commented on doc/releases/release-notes-3.1.rst:
```diff
@@ -87,6 +87,9 @@ Bluetooth
 
   * The :c:macro:bt_conn_index function now takes a `const struct bt_conn`.
 
+  * The `struct bt_gatt_subscribe_params` :c:func:`write` callback
+    function has been replaced with a :c:func:`subscribe` callback
+    function.
```
If we are deprecating the `write` instead of replacing it, this will need to be updated

asbjornsabo replied:
Why?
The write has been replaced by a subscribe, and one is supposed to use the new one.
(It could of course say that the write has been deprecated as well.)

Thalley replied:
Because if it is deprecated, then we did not replace it, but rather `deprecated :c:func:`write` and `added :c:func:`subscribe`. Since both still exist, it has not yet been "replaced".

asbjornsabo replied:
Changed

asbjornsabo commented:
Pushed with changes as discussed in Discord
- documentation change ("deprecated" instead of "replaced")
-  BT_WARN on the deprecated callback
